File: csharp/05_FromPushToPull/PushingBalancesCalculator.cs
using System;
using System.Collections.Generic;
using Common;

namespace FromPushToPull
{
    public class PushingBalancesCalculator : BalancesOfMonthCalculator
    {
        private readonly IList<Transaction> transactions;

        public PushingBalancesCalculator(IList<Transaction> transactions)
        {
            this.transactions = transactions;
        }

        public void FillData(IList<BalancesOfMonth> balancesOfMonthList)
        {
            ValuesOfMonth valuesOfMonth = new ValuesOfMonth();
            int balance = 0;

            foreach (BalancesOfMonth balancesOfMonth in balancesOfMonthList)
            {
                DateTime dateOfMonth = balancesOfMonth.Date;
                IList<Transaction> transactionsOfMonth = TransactionsOfMonth(dateOfMonth);

                valuesOfMonth = CalculateValuesForMonth(balance, dateOfMonth, transactionsOfMonth);
                balance = valuesOfMonth.Balance;

                balancesOfMonth.Balance = valuesOfMonth.Balance;
                balancesOfMonth.AverageBalance = valuesOfMonth.AverageBalance;
            }
        }

        private ValuesOfMonth CalculateValuesForMonth(int precedingBalance, DateTime dateOfMonth, IList<Transaction> transactionsOfMonth)
        {
            ValuesOfMonth valuesOfMonth = new ValuesOfMonth();
            valuesOfMonth.CalculateValues(dateOfMonth, transactionsOfMonth, precedingBalance);
            return valuesOfMonth;
        }

        private IList<Transaction> TransactionsOfMonth(DateTime date)
        {
            IList<Transaction> results = new List<Transaction>();
            foreach (Transaction transaction in transactions)
            {
                DateTime dateOfTransaction = transaction.Date;
                if (AreSameMonthAndYear(date, dateOfTransaction))
                {
                    results.Add(transaction);
                }
            }
            return results;
        }

        private bool AreSameMonthAndYear(DateTime date, DateTime dateOfTransaction)
        {
            return dateOfTransaction.Month == date.Month && dateOfTransaction.Year == date.Year;
        }

    }
}

File: ruby/11_pull/lib/pushing_balances_calculator.rb
require_relative 'values_of_month'

class PushingBalancesCalculator
  attr_reader :transactions
  private :transactions
  def initialize(transactions)
    @transactions = transactions
  end

  def fill_data(balances_of_month_list)
    values_of_month = ValuesOfMonth.new

    balances_of_month_list.each do |balances_of_month|
      date_of_month = balances_of_month.date
      transactions_of_month = transactions_of_month(date_of_month)

      preceding_balance = values_of_month.balance

      values_of_month = ValuesOfMonth.new(date_of_month, transactions_of_month, preceding_balance)

      balances_of_month.balance = values_of_month.balance
      balances_of_month.average_balance = values_of_month.average_balance
    end
  end

  private
  def transactions_of_month(date)
    transactions.select { |transaction|
      are_same_month_and_year(date, transaction.date)
    }
  end

  def are_same_month_and_year(date, date_of_transaction)
    date_of_transaction.month == date.month &&
      date_of_transaction.year == date.year
  end
end
